### PR TITLE
feat: Add TD Ameritrade positions syncing

### LINKED ISSUE #115: Feature Request: TD Ameritrade Sync
**Problem Statement**
It would be super handy if `ticker` could just sync my positions from the [TD Ameritrade API](https://developer.tdameritrade.com/apis) on startup.

**Proposed Solution**
I'd propose reaching out to the [`account-access/apis-get/accounts-0`](https://developer.tdameritrade.com/account-access/apis/get/accounts-0) endpoint (with `fields=positions`) somewhere around [here](https://github.com/achannarasappa/ticker/blob/8a8fd37796b3aae262e69348b80226a7594f647a/internal/cli/cli.go#L139) and just slamming them in a `[]Lot` to be merged with any `lots` from the user's config. I'm slowly working through a PR (dusting off my limited `go` experience) but wanted to get your thoughts on the approach before I send it off.

Initially I'd just expect the user to add a an access token suitable for use with the `Bearer <access token>` Authorization http header in their config as a new yaml option called `td-access-token`. This means the user has to get set up with the TD Ameritrade API themselves ([Getting Started](https://developer.tdameritrade.com/content/getting-started)) and then follow the [Simple Auth for Local Apps](https://developer.tdameritrade.com/content/simple-auth-local-apps) instructions to get their own access token. In the future maybe we could spin up a local webserver on a rando port, but that would require the user get a secure tunnel with `ngrok` or similar since the TD Ameritrade callback apparently has to be https. (There might be another way around that with local ssl certificates but I haven't looked much into it.)

**Alternatives**
I've also thought about automating it externally by operating on my `~/.ticker.yaml` directly but thought something built-in would be cleaner, less moving parts.

### PR DESCRIPTION
**Pull request by mgalgs**
This patch introduces a new config option, `broker-positions-sync`, which specifies broker authentication details in order to sync your positions from your broker automatically.

Currently only TD Ameritrade is supported.

Example:

    # ~/.ticker.yaml
    broker-positions-sync:
      td:
        access-token: "<your_token_here>"

You can get an access token manually with these instructions [1] or with
the `get_td_token.sh` example script in this patch.

Fixes #115

[1] https://developer.tdameritrade.com/content/simple-auth-local-apps

---

This is currently a proof of concept. It's working as is but I've left it as a DRAFT for some additional cleanup if you're interested in merging this feature. The cleanup I'd still like to do is:

- [ ] Move the broker API stuff into a separate file and make it more generic so that other APIs can be easily added.
- [ ] Handle initial oauth token grant as well as token refreshing.
    - Have to decide how we want to handle SSL (see discussion on #115). Options on the table are (1) self-signed cert, or (2) secure tunnel (`ngrok` or similar). I'm leaning towards (2)...
    - Once this is done I'd also be able to rip out `get_td_token.sh`.

But I don't want to do that cleanup work unless you think this feature is worth including in `ticker`. Thanks for taking a look!

### CODE REVIEW

**Comment by cwarden** [comment]
> Have to decide how we want to handle SSL (see discussion on #115). Options on the table are (1) self-signed cert, or (2) secure tunnel (ngrok or similar). I'm leaning towards (2)...

If you want to use a publicly accessible callback endpoint, another option is to have it send the browser to an http localhost URL.  The force cli is an example of an application that used to do this.  It used https://force-cli.herokuapp.com/auth/callback as the callback url, which would use javascript to [get the auth token from the hash and redirect the user](https://github.com/ForceCLI/force-cli/blob/d01845cdd87dcf0431833f8cb526d14f6d6e223f/public/js/callback.js).

It's a bit annoying to have an open source application depend upon the availability and trustworthiness of a web page, but perhaps slightly better than a dependency on a service like ngrok.

File: internal/cli/cli.go [hunk 1]
```diff
@@ -119,6 +121,63 @@ func getReference(config Config, client resty.Client) (Reference, error) {
...
+		Cusip     string `json:"cusip"`
+		Symbol    string `json:"symbol"`
+	} `json:"instrument"`
+	LongQuantity float64 `json:"longQuantity"`
+}
+
+func (p *TDPosition) ToLot() Lot {
+	return Lot{
+		Symbol: p.Instrument.Symbol,
+		UnitCost: p.AveragePrice / p.LongQuantity,
```
Comment (cwarden) [inline internal/cli/cli.go:137]: Looks like `averagePrice` is already the per-share cost.
```suggestion
		UnitCost: p.AveragePrice,
```

Comment (mgalgs) [reply internal/cli/cli.go]: Good catch. Done.

File: get_td_token.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    # cheating... i know...
+    python -c "import urllib.parse; print(urllib.parse.quote('''$1'''))"
+}
+
+uridecode() {
+    python -c "import urllib.parse; print(urllib.parse.unquote('''$1'''))"
+}
+
+get_new_auth() {
+    xdg-open "https://auth.tdameritrade.com/auth?response_type=code&redirect_uri=$(uriencode $TD_REDIRECT_URI)&client_id=$(uriencode $TD_CONSUMER_KEY)%40AMER.OAUTHAP"
```
Comment (cwarden) [inline get_td_token.sh:30]: I know this is an example, but it'd be a good idea to limit the scope of the OAuth token:
```suggestion
    xdg-open "https://auth.tdameritrade.com/auth?response_type=code&redirect_uri=$(uriencode $TD_REDIRECT_URI)&client_id=$(uriencode $TD_CONSUMER_KEY)%40AMER.OAUTHAP&scope=AccountAccess"
```

Comment (mgalgs) [reply get_td_token.sh:30]: Oh this is fantastic, I was actually thinking I needed to look into this. Thanks!

**Comment by cwarden** [comment]
@achannarasappa is modifying `.ticker.yaml`, like the [alpaca integration](https://github.com/achannarasappa/ticker/commit/5c518e51c13980530c312fc6b4e9fe1a1d49b45d), the preferred method of integrating brokerage accounts with ticker?

**Comment by achannarasappa** [comment]
@cwarden thanks for opening this PR and apologies for the lack of feedback on this one.

When it comes to pulling in brokerage specific data, modifying `.ticker.yaml` seems like the right approach to me since position data retrieval seems a bit far afield from the original intent of ticker and these integrations can introduce a good bit complexity.

If you have other ideas on how best to extend ticker I'd be interested to hear those as well but for now, modifying the configuration file seems like the best option.

**Pull request closed without merging**